mmtk/src/object_scanning.rs

```diff
@@ -128,69 +145,96 @@ impl OopIterate for InstanceRefKlass {
+        closure.visit_edge(discovered_addr);
     }
 }
 
 #[allow(unused)]
-fn oop_iterate_slow(oop: Oop, closure: &mut impl EdgeVisitor<OpenJDKEdge>, tls: OpaquePointer) {
+fn oop_iterate_slow<const COMPRESSED: bool, V: EdgeVisitor<E<COMPRESSED>>>(
+    oop: Oop,
+    closure: &mut V,
+    tls: OpaquePointer,
+) {
     unsafe {
-        ((*UPCALLS).scan_object)(closure as *mut _ as _, mem::transmute(oop), tls);
+        CLOSURE.with(|x| *x.get() = closure as *mut V as *mut u8);
+        ((*UPCALLS).scan_object)(
+            mem::transmute(
+                scan_object_fn::<COMPRESSED, V> as *const unsafe extern "C" fn(edge: Address),
+            ),
+            mem::transmute(oop),
+            tls,
+        );
     }
 }
 
-fn oop_iterate(oop: Oop, closure: &mut impl EdgeVisitor<OpenJDKEdge>) {
-    let klass_id = oop.klass.id;
-    debug_assert!(
+fn oop_iterate<const COMPRESSED: bool>(oop: Oop, closure: &mut impl EdgeVisitor<E<COMPRESSED>>) {
+    let klass = oop.klass::<COMPRESSED>();
+    let klass_id = klass.id;
+    assert!(
         klass_id as i32 >= 0 && (klass_id as i32) < 6,
         "Invalid klass-id: {:x} for oop: {:x}",
         klass_id as i32,
         unsafe { mem::transmute::<Oop, ObjectReference>(oop) }
     );
     match klass_id {
         KlassID::Instance => {
-            let instance_klass = unsafe { oop.klass.cast::<InstanceKlass>() };
-            instance_klass.oop_iterate(oop, closure);
+            let instance_klass = unsafe { klass.cast::<InstanceKlass>() };
+            instance_klass.oop_iterate::<COMPRESSED>(oop, closure);
         }
         KlassID::InstanceClassLoader => {
-            let instance_klass = unsafe { oop.klass.cast::<InstanceClassLoaderKlass>() };
-            instance_klass.oop_iterate(oop, closure);
+            let instance_klass = unsafe { klass.cast::<InstanceClassLoaderKlass>() };
+            instance_klass.oop_iterate::<COMPRESSED>(oop, closure);
         }
         KlassID::InstanceMirror => {
-            let instance_klass = unsafe { oop.klass.cast::<InstanceMirrorKlass>() };
-            instance_klass.oop_iterate(oop, closure);
+            let instance_klass = unsafe { klass.cast::<InstanceMirrorKlass>() };
+            instance_klass.oop_iterate::<COMPRESSED>(oop, closure);
         }
         KlassID::ObjArray => {
-            let array_klass = unsafe { oop.klass.cast::<ObjArrayKlass>() };
-            array_klass.oop_iterate(oop, closure);
+            let array_klass = unsafe { klass.cast::<ObjArrayKlass>() };
+            array_klass.oop_iterate::<COMPRESSED>(oop, closure);
         }
         KlassID::TypeArray => {
-            let array_klass = unsafe { oop.klass.cast::<TypeArrayKlass>() };
-            array_klass.oop_iterate(oop, closure);
+            //     let array_klass = unsafe { oop.klass::<COMPRESSED>().cast::<TypeArrayKlass>() };
+            //     array_klass.oop_iterate::<C, COMPRESSED>(oop, closure);
```

Comment:
Are these lines intentionally commented out, or they are just commented out for testing?

Reply:
Removed. This is just a debugging left-over